klippa/src/gvar.rs

```diff
@@ -0,0 +1,172 @@
+//! impl subset() for gvar table
+use crate::{estimate_subset_table_size, Plan, Subset, SubsetError, SubsetFlags};
+
+use write_fonts::{
+    read::{tables::gvar::Gvar, types::GlyphId, FontRef, TopLevelTable},
+    FontBuilder,
+};
+
+const FIXED_HEADER_SIZE: u32 = 20;
+// reference: subset() for gvar table in harfbuzz
+// https://github.com/harfbuzz/harfbuzz/blob/63d09dbefcf7ad9f794ca96445d37b6d8c3c9124/src/hb-ot-var-gvar-table.hh#L411
+impl<'a> Subset for Gvar<'a> {
+    fn subset(
+        &self,
+        plan: &Plan,
+        font: &FontRef,
+        builder: &mut FontBuilder,
+    ) -> Result<(), SubsetError> {
+        let cap = estimate_subset_table_size(font, Gvar::TAG, plan);
+        let mut out: Vec<u8> = Vec::with_capacity(cap);
+
+        //table header: from version to sharedTuplesOffset
+        out.extend_from_slice(self.offset_data().as_bytes().get(0..12).unwrap());
+
+        // glyphCount
+        let num_glyphs = plan.num_output_glyphs.min(0xFFFF) as u16;
+        out.extend_from_slice(&num_glyphs.to_be_bytes());
+
+        let subset_data_size: u32 = plan
+            .new_to_old_gid_list
+            .iter()
+            .filter_map(|x| {
+                if x.0 == GlyphId::NOTDEF
+                    && !plan
+                        .subset_flags
+                        .contains(SubsetFlags::SUBSET_FLAGS_NOTDEF_OUTLINE)
+                {
+                    return None;
+                }
+                self.data_for_gid(x.0).ok().map(|data| data.len() as u32)
+            })
+            .sum();
+
+        // According to the spec: If the short format (Offset16) is used for offsets, the value stored is the offset divided by 2.
+        // So the maximum subset data size that could use short format should be 2 * 0xFFFFu, which is 0x1FFFE
+        let long_offset = if subset_data_size > 0x1FFFE_u32 {
+            1_u16
+        } else {
+            0_u16
+        };
+        // flags
+        out.extend_from_slice(&long_offset.to_be_bytes());
+
+        // calculate sharedTuplesOffset
+        // shared tuples array follow the GlyphVariationData offsets array at the end of the 'gvar' header.
+        let glyph_var_data_offset_size = if long_offset > 0 { 4 } else { 2 };
```

Comment:
In the if long_offset == 1 { .. } else { .. } blocks here and below the code in both branches looks to be nearly identical, differing only by the use of a 2 or 4 width, a divide by 2 for the offset values, and the use of u32/u16 for glyph offsets.

It should be possible to refactor this code to avoid the duplicated code blocks, here's what I'm thinking:
- move the code from here down into a helper function.
- Make it generic and take either u16 or u32 as a type parameter and then write the code in terms of the generic type. This should allow you to avoid switching on long_offset throughout.

Reply:
I've tried my best to generalize them. The checks on long_offset cannot be avoided, put it outside of each for loop is better from a performance perspective